chisel/src/result.rs

```diff
@@ -0,0 +1,223 @@
+//! Chisel execution results structures.
+//! This module implements ChiselResult, RulesetResult, and ModuleResult.
+//! These structures represent execution manifests for chisel modules, and are produced
+//! by ChiselDriver upon completed execution.
+//! ChiselConfig is a nested structure which contains RulesetResults, which in turn contain
+//! ModuleResults. This makes it structurally identical to the ChiselConfig for the driver
+//! execution which produced it.
+//! RulesetResult also implements utilities for writing the resulting Wasm module to file, if the
+//! driver performed any transformations.
+
+use std::error::Error;
+use std::fmt::{self, Display};
+use std::fs::write;
+use std::path::{Path, PathBuf};
+
+use libchisel::{Module, ModuleError};
+
+use ansi_term::Colour::{Green, Red, Yellow};
+
+#[derive(Clone)]
+/// Main result structure returned by ChiselDriver, containing a manifest of modules executed and
+/// exposing methods to write output from translators and creators.
+pub struct ChiselResult(Vec<RulesetResult>);
+
+#[derive(Clone)]
+/// Ruleset result structure.
+pub struct RulesetResult {
+    ruleset_name: String,
+    results: Vec<ModuleResult>,
+    output_path: PathBuf,
+    output_module: Option<Module>,
+}
+
+/// Writer struct for Chisel module output.
+/// Returned by get_writer() in RulesetResult.
+pub enum ChiselOutputWriter {
+    Bin(Module, PathBuf),
+    Hex(Module, PathBuf),
+    #[cfg(feature = "wabt-utils")]
+    Wat(Module, PathBuf),
+}
+
+#[derive(Clone)]
+/// Individual module execution result. Left-hand field is the module name, and left-hand is the
+/// return value.
+pub enum ModuleResult {
+    Creator(String, Result<bool, ModuleError>),
+    Translator(String, Result<bool, ModuleError>),
+    Validator(String, Result<bool, ModuleError>),
+}
+
+impl ChiselResult {
+    pub fn new() -> Self {
+        ChiselResult(Vec::new())
+    }
+
+    pub fn rulesets_mut(&mut self) -> &mut Vec<RulesetResult> {
+        &mut self.0
+    }
+
+    pub fn rulesets(&self) -> &Vec<RulesetResult> {
+        &self.0
+    }
+}
+
+impl RulesetResult {
+    pub fn new(name: String) -> Self {
+        RulesetResult {
+            ruleset_name: name,
+            results: Vec::new(),
+            output_path: PathBuf::new(),
+            output_module: None,
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        self.ruleset_name.as_str()
+    }
+
+    pub fn results_mut(&mut self) -> &mut Vec<ModuleResult> {
+        &mut self.results
+    }
+
+    pub fn set_output_path(&mut self, path: PathBuf) {
+        self.output_path = path;
+    }
+
+    pub fn set_output_module(&mut self, module: Module) {
+        self.output_module = Some(module);
+    }
+
+    /// Get a ChiselOutputWriter from the ruleset and consumes the output module.
+    pub fn get_writer(&mut self, mode: &str) -> Option<ChiselOutputWriter> {
+        if let Some(module) = self.output_module.take() {
```

Comment:
Since `take` replaced `output_module` with `None` it may as well be simpler taking `self` by value and not reference to avoid all this?

The confusing part is that calling `get_writer` successfully make any further call return empty output.